src/common/marlin_server.cpp

```diff
@@ -467,6 +486,282 @@ int marlin_server_print_reheat_ready(void) {
+    _test_sg_sample_on = false;
+    osDelay(100);
+    int32_t length = dir * (stepper.position(Y_AXIS) - _test_start_pos);
+    uint32_t sg_avg = _test_sg_sample_sum / _test_sg_sample_count;
+    f_printf(&_test_fil, "measured length: %d\n", length);
+    f_printf(&_test_fil, "avg stallguard: %d\n\n", sg_avg);
+    f_sync(&_test_fil);
+}
+
+void _test_sg_sample(uint8_t axis, uint16_t sg) {
+    if (!_test_sg_sample_on)
+        return;
+    int32_t pos = stepper.position(X_AXIS);
+    if (_test_fil_ok)
+        f_printf(&_test_fil, "%u %d %d\n", HAL_GetTick() - _test_start_time, pos, sg);
+    _test_sg_sample_count++;
+    _test_sg_sample_sum += sg;
+    //	_dbg("%u %d %d", HAL_GetTick() - _test_start_time, pos, sg);
+}
+
+static void _server_test_loop(void) {
+    switch (marlin_server.test_state) {
+    case mtsIdle:
+        break;
+
+    case mtsStart: {
+        marlin_server.test_type = mttInit;
+        marlin_server.test_mask = (1 << mttInit) | (1 << mttXAxis) | (1 << mttYAxis);
+        marlin_server.flags |= MARLIN_SFLG_EXCMODE; // enter exclusive mode
+        marlin_server.test_state = mtsInit;
+        char *serial_otp = (char *)OTP_SERIAL_NUMBER_ADDR;
+        if (*serial_otp == 0xff) {
+            serial_otp = 0;
+        }
+        char serial[32] = "unknown";
+        char fname[32] = "test_unknown.txt";
+        if (serial_otp) {
+            sprintf(serial, "CZPX%.15s", serial_otp);
+            sprintf(fname, "test_CZPX%.15s.txt", serial_otp);
+        }
+        if (f_open(&_test_fil, fname, FA_WRITE | FA_CREATE_ALWAYS) == FR_OK) {
+            _test_fil_ok = true;
+            f_printf(&_test_fil, "TEST START\n");
+            f_printf(&_test_fil, "printer serial: %s\n\n", serial);
+            f_sync(&_test_fil);
+        } else
+            _test_fil_ok = false;
+        break;
+    }
+
+    case mtsInit:
+        if ((planner.movesplanned() == 0) && (queue.length == 0)) {
+            // TODO: validate system state (temperature, etc)
+            marlin_server_enqueue_gcode("G28");
+            marlin_server.command = MARLIN_CMD_G28;
+            marlin_server.test_state = mtsInit_Home;
+        }
+        break;
+
+    case mtsInit_Home:
+        if (marlin_server.command != MARLIN_CMD_G28) {
+            marlin_server.test_state = mtsXAxis_Start;
+        }
+        break;
+
+    case mtsXAxis_Start:
+        osDelay(100);
+        marlin_server_enqueue_gcode("G28 X");
+        marlin_server.command = MARLIN_CMD_G28;
+        marlin_server.test_state = mtsXAxis_Home;
+        break;
+    case mtsXAxis_Home:
+        if (marlin_server.command != MARLIN_CMD_G28) {
+            marlin_server.test_state = mtsXAxis_Measure_RL_50mms;
+            tmc_sg_mask = 0x01;
+            tmc_sg_sampe_cb = _test_sg_sample;
+            endstops.enable(true);
+            _test_XAxis_start(50, -1);
+        }
+        break;
+    case mtsXAxis_Measure_RL_50mms:
+        if (planner.movesplanned() == 0) {
```

Comment:
I would use some kind of table containing movement data instead repeating code

Reply:
fixed